### DISCUSSION TITLE
RollbackManager improvements

### DISCUSSION THREAD

**Discussion by SnaiperoG3D** [FirstGearGames/FishNet#400] [Approved Changes or Features]
Idk is it overkill or not. I have 2 options to offer.
1) lets say we have team deathmatch setup. 20x20 or 50x50 players. When some player start to fire, rollbackmanager will rollback all subscribed colliders. Each player can have up to 10 colliders. As as much players shooting in same time, as more performance it will eat. So what i did in my case. First split each team hitbox in separate layers for collision matrix: RedHitbox and BlueHitbox (could be more than 2 teams). Now we need to check for Raycasts and Overlaps only opposite layer. But there is still a problem with rollbacks that rollbackmanager will rollback all 40x10 or 100x10 colliders (depends of what colliders count per player). So what i did. Ive added to ColliderRollback.cs an int colliderlayer var which holds a layer value of all colliders of this ColliderRollback and added an arg LayerMask? to RollbackManager_OnRollback method, where i do check, is arg layermask do have ColliderRollback layer in it, or not. If not, then return. Now my RollbackManager will rollback only colliders with needed layer.
Here is a code. 
ColliderRollback.cs changes: https://pastebin.com/ze6ZKnPr
RollbackManager.cs changes: https://pastebin.com/TpqJgGuT

Idk will it improve performance or not yet, i didnt do any stress tests yet, but in theory in this setup it should be twice cheaper.

2) lets say we have big open world map with a lot of players 50+ like DayZ, enemies, animals and all stuff that could have ColliderRollback script on it. In that case we have same problem, that when player from one side of the map will shoot, colliders on whole map will be rolled back. I guess for that case RollbackManager should have something like SceneCondition, or DistanceCondition, or maybe when player shoot, dev should do boxoverlap with lenght of raycast and width of max rollback time * player movespeed, and all RollbackCollider.cs what be overlapped will be feeded to RollbackManager to do rollback. 

Idk are these ideas is overkill or not.

**Comment by FirstGearGames** [comment]
Rollbacks really are not that expensive for even a hundred colliders. There are ideas I've been considering to improve the performance regardless but they are a bit low priority at the moment. I wanted to update you as I know it's been awhile, and this is still planned, but it can take a bit for me to get to it.

**Comment by FirstGearGames** [comment]
Rollback improvements added in a few releases ago. You can now use bounding boxes or scene specific rollbacks as well.
